src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,145 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final int DIVIDE_BY_TWO = 2;
+    private static final int INDEX_BY_ZERO = 0;
+    private static final int INDEX_BY_ONE = 1;
+    private static final int EMPTY = 0;
+    private T[] array;
+    private T[] copiedArray;
+    private int size;
+
+    public ArrayList() {
+        array = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (array.length == size) {
+            array = grow();
+        }
+        array[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < INDEX_BY_ZERO) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "You cannot add element by negative index");
+        }
+        if (array.length == size + INDEX_BY_ONE) {
+            array = grow();
+        }
+        if (index < array.length && index < size) {
+            copiedArray = (T[]) new Object[array.length];
+            System.arraycopy(array, INDEX_BY_ZERO, copiedArray, INDEX_BY_ZERO, index);
+            copiedArray[index] = value;
+            System.arraycopy(array, index, copiedArray, index + INDEX_BY_ONE, size + INDEX_BY_ONE);
+            System.arraycopy(copiedArray, INDEX_BY_ZERO, array, INDEX_BY_ZERO, size + INDEX_BY_ONE);
+            size++;
+            return;
+        } else if (index > size) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "You cannot add element by not existent index");
+        }
+        array[size] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < size && index >= INDEX_BY_ZERO) {
+            return array[index];
+        }
+        throw new ArrayListIndexOutOfBoundsException("There is not element by index: " + index);
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= size) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "You cannot add element by not existent index");
+        } else if (index < INDEX_BY_ZERO) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "You cannot add element by negative index");
+        }
```

Comment:
here just call your checker method - `checkIndexInBounds`
The message will be generic but that's not a problem, inlucde size and index to the message and it will be informative enough for the user of your class to figure out what is wrong